src/syntaxtree.c

```diff
@@ -31,50 +31,71 @@
 #include <ctype.h>
 #include <stdarg.h>
 
+#define MAX_ARITY 4
+
 typedef Obj (*CompileFuncT)(Obj node, Expr expr);
 typedef Obj (*CompileArgT)(Expr expr);
+typedef Expr (*CodeFuncT)(Obj node);
+typedef Expr (*CodeArgT)(Obj node);
+
 
 typedef struct {
     const Char * argname;
     CompileArgT  argcomp;
+    CodeArgT     argcode;
 } ArgT;
 
 typedef struct {
     UInt1        tnum;
     CompileFuncT compile;
+    CodeFuncT    code;
     const Char * name;
-    UInt1        arity;
-    ArgT         args[4]; /* The maximum compiler arity is currently 4 */
+    UInt         arity;
+    ArgT args[MAX_ARITY]; /* The maximum compiler arity is currently 4 */
 } CompilerT;
 
 // We put compilers for statements and expressions into the same static array,
 // assuming that the set of their respective ids are disjoint.
 static const CompilerT Compilers[];
 #define COMPILER_ARITY(...) (sizeof((ArgT[]){ __VA_ARGS__ }) / sizeof(ArgT))
-#define COMPILER_GENERIC(tnum, compiler, ...)                                \
+#define COMPILER_GENERIC(tnum, compiler, coder, ...)                         \
     [tnum] = {                                                               \
-        tnum, compiler, #tnum, COMPILER_ARITY(__VA_ARGS__), { __VA_ARGS__ }  \
+        tnum,           compiler, coder, #tnum, COMPILER_ARITY(__VA_ARGS__), \
+        { __VA_ARGS__ }                                                      \
     }
 
-#define COMPILER(tnum, compiler) COMPILER_GENERIC(tnum, compiler)
+#define COMPILER(tnum, compiler, coder)                                      \
+    COMPILER_GENERIC(tnum, compiler, coder)
 
 #define COMPILER_(tnum, ...)                                                 \
-    COMPILER_GENERIC(tnum, SyntaxTreeDefaultCompiler, __VA_ARGS__)
+    COMPILER_GENERIC(tnum, SyntaxTreeDefaultCompiler,                        \
+                     SyntaxTreeDefaultCoder, __VA_ARGS__)
 
-#define ARG(name, func)                                                      \
+#define ARG(name, func, coder)                                               \
     {                                                                        \
-        name, func                                                           \
+        name, func, coder                                                    \
     }
-#define ARG_(name) ARG(name, SyntaxTreeCompiler)
+#define ARG_(name) ARG(name, SyntaxTreeCompiler, SyntaxTreeDefaultCoder)
 
-#define ARGS(name) ARG(name, 0)
+// use ARGS as last "arg" which collect all the following args together
+#define ARGS(name) ARG(name, 0, 0)
 
 
 static Obj typeStrings;
 
 
 static Obj SyntaxTreeFunc(Obj result, Obj func);
 
+static UInt FIND_TNUM_FROM_STRING(char * tnum)
+{
+    for (UInt i = 0; i <= LAST_EXPR_TNUM; i++) {
```

Comment:
Safer would be this:
```suggestion
    for (UInt i = 0; i < ARRAY_SIZE(Compilers); i++) {
```

Reply:
This does not compile, as `Compilers` only gets initialized afterwards. We would have to move the function further down and forward-declare it.